If all packages are opam-provided, an invalid lockfile is created

Is opam-monorepo so powerful that it can create a lockfile that it can't pull? Turns out, yes :/

```
 $ cat src.opam
opam-version: "2.0"
depends: ["lwt"]
x-opam-monorepo-opam-provided: ["lwt"]
 $ opam monorepo lock        
==> Using 1 locally scanned package as the target.
==> Found 16 opam dependencies for the target package.
==> Querying opam database for their metadata and Dune compatibility.
==> Calculating exact pins for each of them.
==> Wrote lockfile with 0 entries to src.opam.locked. You can now run opam monorepo pull to fetch their sources.
 $ opam monorepo pull
==> Using lockfile src.opam.locked
opam-monorepo: [ERROR] Missing x-opam-monorepo-duniverse-dirs field in opam-monorepo lockfile /tmp/bug/src.opam.locked
```

The reason is that the lockfile does not, indeed, contain this field.

After a light investigation, it seems that it's due to how the opam libs render empty lists:

```
utop # OpamFile.OPAM.empty
  |> OpamFile.OPAM.with_extensions
      (OpamStd.String.Map.singleton "x-field"
        (OpamTypesBase.nullify_pos (OpamParserTypes.FullPos.List (OpamTypesBase.nullify_pos []))))
  |> OpamFile.OPAM.write_to_string;;
- : string = "opam-version: \"2.0\"\n"
```

So from the point of view of `opam-monorepo`, I think that it's necessary to handle the absence of this field as equivalent to an empty list. A special empty value is possible too I guess but since lockfiles are versioned I don't think there can be a confusion between "no field" and "an empty list".

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -14,7 +14,7 @@
 - Enable locking of packages with depexts even with uninitialized system
   package manager state (#322, @Leonidas-from-XIV)
 - Fix a bug where `pull` would crash if the lock file contained no package to
-  vendor (#<PR_NUMBER>, @NathanReb)
+  vendor (#321, @NathanReb)
 
 ### Removed
 

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -13,6 +13,8 @@
 
 - Enable locking of packages with depexts even with uninitialized system
   package manager state (#322, @Leonidas-from-XIV)
+- Fix a bug where `pull` would crash if the lock file contained no package to
+  vendor (#<PR_NUMBER>, @NathanReb)
 
 ### Removed
 

--- a/lib/lockfile.ml
+++ b/lib/lockfile.ml
@@ -3,12 +3,13 @@ open Import
 module Extra_field = struct
   include Opam.Extra_field
 
-  let get ?file t opam =
+  let get ?file ?default t opam =
     let open Result.O in
     let* value_opt = get t opam in
-    match value_opt with
-    | Some result -> Ok result
-    | None ->
+    match (value_opt, default) with
+    | Some result, _ -> Ok result
+    | None, Some default -> Ok default
+    | None, None ->
         let file_suffix_opt = Option.map ~f:(Printf.sprintf " %s") file in
         let file_suffix = Option.value ~default:"" file_suffix_opt in
         Error
@@ -87,6 +88,8 @@ module Root_packages = struct
 
   let field =
     Extra_field.make ~name:"root-packages" ~to_opam_value ~from_opam_value
+
+  let get ?file opam = Extra_field.get ?file field opam
 end
 
 module Depends = struct
@@ -236,6 +239,9 @@ module Duniverse_dirs = struct
 
   let field =
     Extra_field.make ~name:"duniverse-dirs" ~to_opam_value ~from_opam_value
+
+  let default = OpamUrl.Map.empty
+  let get ?file opam = Extra_field.get ?file ~default field opam
 end
 
 module Depexts = struct
@@ -362,10 +368,10 @@ let from_opam ~opam_monorepo_cwd ?file opam =
   let open Result.O in
   let* version = Extra_field.get ?file Version.field opam in
   let* () = Version.compatible version in
-  let* root_packages = Extra_field.get ?file Root_packages.field opam in
+  let* root_packages = Root_packages.get ?file opam in
   let* depends = Depends.from_filtered_formula (OpamFile.OPAM.depends opam) in
   let pin_depends = OpamFile.OPAM.pin_depends opam in
-  let* duniverse_dirs = Extra_field.get ?file Duniverse_dirs.field opam in
+  let* duniverse_dirs = Duniverse_dirs.get ?file opam in
   let depexts = OpamFile.OPAM.depexts opam in
   let* source_config = Source_opam_config.get ~opam_monorepo_cwd opam in
   Ok

--- a/test/bin/empty-duniverse.t/empty-duniverse.opam
+++ b/test/bin/empty-duniverse.t/empty-duniverse.opam
@@ -1,0 +1,8 @@
+opam-version: "2.0"
+depends: [
+  "ocaml"
+  "dune"
+]
+x-opam-monorepo-opam-repositories: [
+  "file://$OPAM_MONOREPO_CWD/minimal-repo"
+]

--- a/test/bin/empty-duniverse.t/run.t
+++ b/test/bin/empty-duniverse.t/run.t
@@ -1,0 +1,30 @@
+One should be able to generate a lock file where no dependencies need to
+be vendored and that should result in an empty duniverse.
+That can happen if one has no external deps besides 'dune' and 'ocaml'
+or if those are marked as 'opam-provided' for example.
+
+Here our local package only depends on 'ocaml' and 'dune':
+
+  $ opam show --just-file -fdepends ./empty-duniverse.opam
+  ocaml, dune
+
+We should be able to successfully lock:
+
+  $ gen-minimal-repo
+  $ opam-monorepo lock > /dev/null
+
+And the lock file should not contain anything to vendor:
+
+  $ opam show --just-file -fdepends ./empty-duniverse.opam.locked | grep "\?vendor"
+  [1]
+
+Finally, we should be able to pull this lock file, the tool will inform us that
+there is nothing to pull and will not create a duniverse:
+
+  $ opam-monorepo pull
+  ==> Using lockfile $TESTCASE_ROOT/empty-duniverse.opam.locked
+  opam-monorepo: [ERROR] Missing x-opam-monorepo-duniverse-dirs field in opam-monorepo lockfile $TESTCASE_ROOT/empty-duniverse.opam.locked
+  [1]
+  $ ls duniverse
+  ls: cannot access 'duniverse': No such file or directory
+  [2]

--- a/test/bin/empty-duniverse.t/run.t
+++ b/test/bin/empty-duniverse.t/run.t
@@ -24,6 +24,4 @@ there is nothing to pull and will not create a duniverse:
   $ opam-monorepo pull
   ==> Using lockfile $TESTCASE_ROOT/empty-duniverse.opam.locked
   ==> No dependencies to pull, there's nothing to be done here!
-  $ ls duniverse
-  ls: cannot access 'duniverse': No such file or directory
-  [2]
+  $ find . -type d -name 'duniverse'

--- a/test/bin/empty-duniverse.t/run.t
+++ b/test/bin/empty-duniverse.t/run.t
@@ -23,8 +23,7 @@ there is nothing to pull and will not create a duniverse:
 
   $ opam-monorepo pull
   ==> Using lockfile $TESTCASE_ROOT/empty-duniverse.opam.locked
-  opam-monorepo: [ERROR] Missing x-opam-monorepo-duniverse-dirs field in opam-monorepo lockfile $TESTCASE_ROOT/empty-duniverse.opam.locked
-  [1]
+  ==> No dependencies to pull, there's nothing to be done here!
   $ ls duniverse
   ls: cannot access 'duniverse': No such file or directory
   [2]

